addons/emoji-picker/userscript.js

```diff
@@ -0,0 +1,551 @@
+  const unicodeEmojis = [
+    "⌚️",
+    "⏰",
+    "⏱️",
+    "⏲️",
+    "⌨️",
+    "☎️",
+    "⌛️",
+    "⚖️",
+    "⚙️",
+    "✉️",
+    "✂️",
+    "✒️",
+    "☂️",
+    "✏️",
+    "☕️",
+    "♟️",
+    "⚰️",
+    "⚱️",
+    "⛏️",
+    "⚔️",
+    "⚒️",
+    "⛓️",
+    "⚗️",
+    "⛑️",
+    "br",
+    "⭐️",
+    "✨",
+    "⚡️",
+    "☄️",
+    "☀️",
+    "⛅️",
+    "☁️",
+    "⛈️",
+    "⛄️",
+    "❄️",
+    "☔️",
+    "☘️",
+    "br",
+    "☹️",
+    "✌️",
+    "☝️",
+    "✍️",
+    "✋",
+    "☺️",
+    "br",
+    "⚽️",
+    "⚾️",
+    "⛳️",
+    "⛹️",
+    "⛷️",
+    "⛸️",
+    "br",
+    "✈️",
+    "⛵️",
+    "Ⓜ️",
+    "⚓️",
+    "⛽️",
+    "⛲️",
+    "⛺️",
+    "⛪️",
+    "⛰️",
+    "⛱️",
+    "⛴️",
+    "⛩️",
+    "♨️",
+    "br",
+    "❗️",
+    "❕",
+    "❓",
+    "❔",
+    "©️",
+    "®️",
+    "‼️",
+    "⁉️",
+    "™️",
+    "➕",
+    "➖",
+    "➗",
+    "✖️",
+    "⛔",
+    "⭕",
+    "❌",
+    "✔️",
+    "〰️",
+    "〽️",
+    "⚠️",
+    "br",
+    "☑️",
+    "✅",
+    "❎",
+    "▶️",
+    "⏩",
+    "⏪",
+    "⏫",
+    "⏬",
+    "ℹ️",
+    "⏭️",
+    "⏮️",
+    "⏯️",
+    "⏏️",
+    "◀️",
+    "➡️",
+    "⬅️",
+    "⬆️",
+    "⬇️",
+    "↗️",
+    "↘️",
+    "↙️",
+    "↖️",
+    "↪️",
+    "↩️",
+    "⤴️",
+    "⤵️",
+    "✳️",
+    "✴️",
+    "❇️",
+    "㊗️",
+    "㊙️",
+    "br",
+    "⚕️",
+    "☦️",
+    "♾️",
+    "⚛️",
+    "⛎️",
+    "✝️",
+    "☪️",
+    "☮️",
+    "☯️",
+    "☸️",
+    "♈️",
+    "♉️",
+    "♊️",
+    "♋️",
+    "♌️",
+    "♍️",
+    "♎️",
+    "♏️",
+    "♐️",
+    "♑️",
+    "♒️",
+    "♓️",
+    "br",
+    "❤️",
+    "❣️",
+    "♠️",
+    "♣️",
+    "♥️",
+    "♦️",
+    "♀️",
+    "♂️",
+    "♻️",
+    "☢️",
+    "☣️",
+    "⚜️",
+    "➰️",
+    "➿️",
+    "☠️",
+    "br",
+    "⬛️",
+    "⬜️",
+    "⚪️",
+    "⚫️",
+    "▪️",
+    "▫️",
+    "◻️",
+    "◼️",
+    "◽️",
+    "◾️",
+  ];
+
+  const emojiPickerOffset = addon.tab.clientVersion === "scratchr2" ? 30 : 52;
+
+  //Functions
+
+  const setEmojiPickerPos = function () {
+    emojiPicker.style.top = emojiPickerOffset + "px";
+    //scratchr2 makes the body and root <html>'s height value the size of the screen somehow so I have to do this
+    let realDocumentBody =
+      addon.tab.clientVersion === "scratchr2" ? document.querySelector("#pagewrapper") : document.body;
+    if (emojiPicker.getBoundingClientRect().bottom > realDocumentBody.getBoundingClientRect().bottom - 48) {
+      //Emoji picker may be partially hidden, move up
+      emojiPicker.style.top = -emojiPicker.getBoundingClientRect().height + "px";
+    }
+  };
+
+  const showEmojiPicker = function (event) {
+    if (!event.target.classList.contains("sa-emoji-button")) return; //Only attempt to show when clicking button, not picker
+    unicodeContainer.style.display = "none";
+    pickerDivider.style.display = "none";
+
+    setSeeMoreText();
+
+    this.appendChild(emojiPicker);
+
+    setEmojiPickerPos();
+  };
+
+  const insertTextToTextArea = function (insertText, textBox) {
```

Comment:
text-field-edit?

Reply:
(Apple is referring to the library that we already ship for image-uploader)

Reply:
I didn't know.

Reply:
Actually, my function inserts spaces if there aren't already next to the added text. (I don't know how to do that with the library insert function)